cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/TransformRunner.java

```diff
@@ -0,0 +1,176 @@
+    MapReduceTaskContext<Object, Object> context,
+    PipelinePluginInstantiator pluginInstantiator,
+    Metrics metrics,
+    Map<String, Map<String, String>> runtimeArgs,
+    PipelinePhase pipelinePhase
+  ) throws Exception;
+
+  // this is needed because we need to write to the context differently depending on the number of outputs
+  private OutputWriter<Object, Object> getSinkWriter(MapReduceTaskContext<Object, Object> context,
+                                                     PipelinePhase pipelinePhase,
+                                                     Configuration hConf) {
+    Set<StageInfo> aggregators = pipelinePhase.getStagesOfType(BatchAggregator.PLUGIN_TYPE);
+    if (!aggregators.isEmpty()) {
+      String aggregatorName = aggregators.iterator().next().getName();
+      if (pipelinePhase.getSinks().contains(aggregatorName)) {
+        return new SingleOutputWriter<>(context);
+      }
+    }
+
+    String sinkOutputsStr = hConf.get(ETLMapReduce.SINK_OUTPUTS_KEY);
+
+    // should never happen, this is set in beforeSubmit
+    Preconditions.checkNotNull(sinkOutputsStr, "Sink outputs not found in Hadoop conf.");
+    Map<String, SinkOutput> sinkOutputs = GSON.fromJson(sinkOutputsStr, ETLMapReduce.SINK_OUTPUTS_TYPE);
+    return hasSingleOutput(pipelinePhase.getStagesOfType(Transform.PLUGIN_TYPE), sinkOutputs) ?
+      new SingleOutputWriter<>(context) : new MultiOutputWriter<>(context, sinkOutputs);
+  }
+
+  private boolean hasSingleOutput(Set<StageInfo> transformInfos, Map<String, SinkOutput> sinkOutputs) {
+    // if there are any error datasets, we know we have at least one sink, and one error dataset
+    for (StageInfo info : transformInfos) {
+      if (info.getErrorDatasetName() != null) {
+        return false;
+      }
+    }
+    // if no error datasets, check if we have more than one sink
+    Set<String> allOutputs = new HashSet<>();
+
+    for (SinkOutput sinkOutput : sinkOutputs.values()) {
+      if (sinkOutput.getErrorDatasetName() != null) {
+        return false;
+      }
+      allOutputs.addAll(sinkOutput.getSinkOutputs());
+    }
+    return allOutputs.size() == 1;
+  }
+
+  public void transform(KEY key, VALUE value) {
+    try {
+      KeyValue<KEY, VALUE> input = new KeyValue<>(key, value);
+      TransformResponse transformResponse = transformExecutor.runOneIteration(input);
+      for (Map.Entry<String, Collection<Object>> transformedEntry : transformResponse.getSinksResults().entrySet()) {
+        for (Object transformedRecord : transformedEntry.getValue()) {
+          outputWriter.write(transformedEntry.getKey(), (KeyValue<Object, Object>) transformedRecord);
+        }
+      }
+
+      for (Map.Entry<String, Collection<InvalidEntry<Object>>> errorEntries :
+        transformResponse.getMapTransformIdToErrorEmitter().entrySet()) {
+
+        if (transformsWithoutErrorDataset.contains(errorEntries.getKey())) {
+          continue;
```

Comment:
er yeah actually now i get it, it's like you said, used to avoid logging the same error multiple times.